[CodeComplete] A couple of contextual type handling tweaks

hamishknight opened the pull request:
Found while cleaning up the constraint system's handling of ReturnStmts, improve the handling of contextual types in a couple of cases.

hamishknight commented:
@swift-ci please test

ahoppen commented on lib/IDE/PostfixCompletion.cpp:
```diff
@@ -189,23 +189,24 @@ void PostfixCompletionCallback::sawSolutionImpl(
-      !ParentExpr && CS.getContextualTypePurpose(CompletionExpr) != CTP_Unused;
+    ExpectsNonVoid = !ExpectedTy->isVoid();
+  } else {
+    ExpectsNonVoid |= !ParentExpr &&
```
I don’t know when it would happen but if expected type is non-void but it’s unused by the constraint system, we should still have `ExpectsNonVoid = false`, I think. Did you intentionally change it? Same for the check in the loop below.

hamishknight replied:
> if expected type is non-void but it’s unused by the constraint system, we should still have `ExpectsNonVoid = false`, I think

Changing this means that for e.g the following:

```
  let _ = MyStruct4 {
    1
  } #^INIT_FALLBACK_1^#
// INIT_FALLBACK_1: Begin completions, 3 items
// INIT_FALLBACK_1-DAG: Keyword[self]/CurrNominal:          .self[#MyStruct4<Int>#]; name=self
// INIT_FALLBACK_1-DAG: Decl[InstanceMethod]/CurrNominal/TypeRelation[Invalid]: .testStructMethod()[#Void#]; name=testStructMethod()
```

`testStructMethod()` is no longer marked as having an `Invalid` type relation, which I guess is reasonable? It's not invalid, although it's also not particularly useful either.

ahoppen replied:
Yes, I think that’s reasonable. If `ExpectedTy` is null, we don’t know what the expected type is and thus we can’t tell that it’s not `Void`.

What I was talking about, though, is that `ExpectsNonVoid` should be `false` if `ExpectedTy` is eg. `Int` but the contextual type purpose of the code completion expression is unused.

hamishknight replied:
How's this?

ahoppen replied:
Almost. 

```cpp
  auto isResultUsed = [&]() {
    if (!ParentExpr &&
        CS.getContextualTypePurpose(CompletionExpr) != CTP_Unused) {
      return true;
    }
    for (auto SAT : S.targets) {
      if (SAT.second.getAsExpr() == CompletionExpr)
        return SAT.second.getExprContextualTypePurpose() != CTP_Unused;
    }
    // No contextual type purpose, assume it's unused.
    return false;
  };

  bool ExpectsNonVoid = (ExpectedTy && !ExpectedTy->isVoid()) || isResultUsed();
```

That also maintains the current behavior.

hamishknight commented:
@swift-ci please test

hamishknight committed:
[CodeComplete] Improve `ExpectsNonVoid` computation

If we know we have a Void expected type, don't set
`ExpectsNonVoid` to `true`.

hamishknight committed:
[CodeComplete] Avoid using an unresolved contextual type

This can occur for e.g multi-statement closure
returns (and soon for singe-expression closures
too), so fall back to evaluating the expression
type if it's not fully resolved.

hamishknight commented:
@swift-ci please test

ahoppen commented on test/IDE/complete_single_expression_return.swift:
```diff
@@ -128,6 +128,55 @@ struct TestSingleExprClosureGlobal {
+      return self.#^TestExplicitSingleExprClosureBinding^#
+    }
+  }
+// FIXME: Because we have an explicit return, and no expected type, we shouldn't suggest Void.
```
Actually, I think suggesting values that return `Void` after `return` makes sense if the closure has a `Void` return type. For example `return callback()` is a common pattern for early returns out of closures and as a shorthand of `callback(); return`

So, I would just remove the FIXME

hamishknight replied:
Right, but in this case there's no contextual type. If there was an explicit `Void` contextual type, we'd allow it

hamishknight replied:
We could carve out an exception for `CTP_ReturnStmt` if we really wanted this behavior though

ahoppen replied:
I think what I’m trying to say is that I don’t really care about this case because you can argue either way and I would just remove the FIXME. But if you want to keep it, that’s good with me as well.

hamishknight replied:
Going to keep it for now since this is going to change anyway with my constraint system patch. I should also note that not suggesting Void is also the behavior we currently have for multi-statement closure returns. I don't really feel too strongly about it either way (and am happy to carve out an exception for returns if we want that), but I think we should at least be consistent there

hamishknight replied:
Changing in https://github.com/apple/swift/pull/71272
